grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/HttpStreamReader.java

```diff
@@ -20,202 +20,105 @@
+            if (decompressor == null) {
+                transportStatusListener.transportReportStatus(Status.INTERNAL.withDescription(
+                        "Can't find decompressor for " + grpcEncoding));
+                return;
             }
-            requestHttpFrame();
-            return;
-        }
-        final HttpData data = (HttpData) obj;
-        try {
-            deframer.deframe(data, false);
-        } catch (Throwable cause) {
             try {
-                transportStatusListener.transportReportStatus(GrpcStatus.fromThrowable(cause));
-                return;
-            } finally {
-                deframer.close();
+                decompressor(ForwardingDecompressor.forGrpc(decompressor));
+            } catch (Throwable t) {
+                transportStatusListener.transportReportStatus(GrpcStatus.fromThrowable(t));
             }
         }
-        requestHttpFrame();
     }
 
     @Override
-    public void onError(Throwable cause) {
-        if (cancelled) {
-            return;
+    protected void processTrailers(HttpHeaders headers, HttpDeframerOutput<DeframedMessage> out) {
+        final String grpcStatus = headers.get(GrpcHeaderNames.GRPC_STATUS);
+        if (grpcStatus != null) {
+            whenConsumed().thenRun(() -> GrpcStatus.reportStatus(headers, this, transportStatusListener));
         }
-        transportStatusListener.transportReportStatus(GrpcStatus.fromThrowable(cause));
     }
 
     @Override
-    public void onComplete() {
-        if (cancelled) {
-            return;
-        }
-        closeDeframer();
+    protected void processOnError(Throwable cause) {
+        transportStatusListener.transportReportStatus(GrpcStatus.fromThrowable(cause));
     }
 
+    /**
+     * Cancel this stream and prevents further subscription.
+     */
     public void cancel() {
-        cancelled = true;
-        if (subscription != null) {
-            subscription.cancel();
-        }
-        if (!deframer.isClosed()) {
-            deframer.close();
-        }
-    }
-
-    void closeDeframer() {
-        // closeDeframer() could be called when deframer.isClosing() due to a race condition like the following:
-        //
-        // 1) HttpStreamReader received all data from publisher and added them to unprocessed of deframer.
-        // 2) A gRPC client does not request next messages yet, so deframer still has unprocessedBytes and
-        //    is not stalled.
-        // 3) HttpStreamReader receives onCompleted signal and closes deframer.
-        // 4) A gRPC client requests a message and the received message contains trailers,
-        //    so ArmeriaClientCall tries to close deframer.
-        if (!deframer.isClosing() && !deframer.isClosed()) {
-            deframer.deframe(HttpData.empty(), true);
-            deframer.closeWhenComplete();
-        }
+       abort(CancelledSubscriptionException.get());
```

Comment:
It's a bit weird that aborting with `CancelledSubscriptionException`.
How about just `aborting`?

Reply:
This was my considering point. If it is no harm to call `abort()` without `CancelledSubscriptionException`, let me address it.